common/elasticsearch/interfaces.go

```diff
@@ -0,0 +1,225 @@
+		// Next implements a BackoffFunc.
+		Next(retry int) (time.Duration, bool)
+	}
+
+	// GenericBulkBeforeFunc defines the signature of callbacks that are executed
+	// before a commit to Elasticsearch.
+	GenericBulkBeforeFunc func(executionId int64, requests []GenericBulkableRequest)
+
+	// GenericBulkAfterFunc defines the signature of callbacks that are executed
+	// after a commit to Elasticsearch. The err parameter signals an error.
+	GenericBulkAfterFunc func(executionId int64, requests []GenericBulkableRequest, response *GenericBulkResponse, err *GenericError)
+
+	// IsRecordValidFilter is a function to filter visibility records
+	IsRecordValidFilter func(rec *p.InternalVisibilityWorkflowExecutionInfo) bool
+
+	// GenericBulkableRequest is a generic interface to bulkable requests.
+	GenericBulkableRequest interface {
+		fmt.Stringer
+		Source() ([]string, error)
+	}
+
+	// GenericBulkableAddRequest a struct to hold a bulk request
+	GenericBulkableAddRequest struct {
+		Index       string
+		Type        string
+		Id          string
+		VersionType string
+		Version     int64
+		// true means it's delete, otherwise it's a index request
+		IsDelete bool
+		// should be nil if IsDelete is true
+		Doc interface{}
+	}
+
+	// GenericBulkResponse is generic struct of bulk response
+	GenericBulkResponse struct {
+		Took   int                                   `json:"took,omitempty"`
+		Errors bool                                  `json:"errors,omitempty"`
+		Items  []map[string]*GenericBulkResponseItem `json:"items,omitempty"`
+	}
+
+	// GenericError encapsulates error status and details returned from Elasticsearch.
+	GenericError struct {
+		Status  int   `json:"status"`
+		Details error `json:"error,omitempty"`
+	}
+
+	// GenericBulkResponseItem is the result of a single bulk request.
+	GenericBulkResponseItem struct {
+		Index         string `json:"_index,omitempty"`
+		Type          string `json:"_type,omitempty"`
+		Id            string `json:"_id,omitempty"`
+		Version       int64  `json:"_version,omitempty"`
+		Result        string `json:"result,omitempty"`
+		SeqNo         int64  `json:"_seq_no,omitempty"`
+		PrimaryTerm   int64  `json:"_primary_term,omitempty"`
+		Status        int    `json:"status,omitempty"`
+		ForcedRefresh bool   `json:"forced_refresh,omitempty"`
+		// the error details
+		Error interface{}
+	}
+
+	// VisibilityRecord is a struct of doc for deserialization
+	VisibilityRecord struct {
+		WorkflowID    string
+		RunID         string
+		WorkflowType  string
+		StartTime     int64
+		ExecutionTime int64
+		CloseTime     int64
+		CloseStatus   workflow.WorkflowExecutionCloseStatus
+		HistoryLength int64
+		Memo          []byte
+		Encoding      string
+		TaskList      string
+		Attr          map[string]interface{}
+	}
+)
+
+// ExponentialBackoff implements the simple exponential backoff described by
+// Douglas Thain at http://dthain.blogspot.de/2009/02/exponential-backoff-in-distributed.html.
+type ExponentialBackoff struct {
```

Comment:
Technically yes, but not that simple. I simply copied it from the elasticsearch library, and make an interface 
```
	GenericBackoff interface {
		// Next implements a BackoffFunc.
		Next(retry int) (time.Duration, bool)
	}
```
The common/backoff one is more complicated than that. And to lower the risk of changing behavior, since there are too many changes, I recommend to keep this for now. I could leave a TODO and an issue for this and improve it later. How does that sound?

Reply:
I added a TODO comment and an issue for that.